### PR TITLE
small fixes in ImageSet

### PR DESCRIPTION
**Pull request by ranjithkumar007**

### CODE REVIEW

**Comment by ranjithkumar007** [comment]
@nishnik , can you have a look?

File: symengine/sets.cpp [hunk 1]
```diff
@@ -15,8 +15,8 @@ Interval::Interval(const RCP<const Number> &start, const RCP<const Number> &end,
 }
 
 bool Interval::is_canonical(const RCP<const Number> &s,
-                            const RCP<const Number> &e, bool left_open,
-                            bool right_open)
+                            const RCP<const Number> &e, const bool &left_open,
```
Comment (isuruf) [inline symengine/sets.cpp]: there's no need for this change.

File: symengine/sets.cpp [hunk 2]
```diff
@@ -766,7 +766,10 @@ bool ImageSet::__eq__(const Basic &o) const
 {
     if (is_a<ImageSet>(o)) {
         const ImageSet &other = down_cast<const ImageSet &>(o);
-        return unified_eq(sym_, other.sym_) and unified_eq(expr_, other.expr_)
+        map_basic_basic d;
+        d[sym_]
+            = other.sym_; // ImageSet(x,x**2,base) == ImageSet(y, y**2, base)
```
Comment (isuruf) [inline symengine/sets.cpp]: Can you add a test for this?

File: symengine/sets.cpp (hunk 1 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: making a const ref. is encouraged in such cases right?

Comment (isuruf) [inline symengine/sets.cpp]: For integer types there's no difference and just `bool` improves readability.

**Comment by ranjithkumar007** [comment]
ping @isuruf

File: symengine/sets.cpp [hunk 3]
```diff
@@ -766,7 +766,10 @@ bool ImageSet::__eq__(const Basic &o) const
...
+        return unified_eq(expr_->subs(d), other.expr_)
```
Comment (isuruf) [inline symengine/sets.cpp]: This breaks the hash contract.

Comment (ranjithkumar007) [inline symengine/sets.cpp]: breaking hash contract means two diff objects with different hashes, are considered as equal?

Comment (isuruf) [inline symengine/sets.cpp]: Yes, consider `ImageSet(x,x**2,base)` and `ImageSet(y, y**2, base)`

Comment (ranjithkumar007) [inline symengine/sets.cpp]: How to fix this?

Comment (isuruf) [inline symengine/sets.cpp]: Go back to how it was?

Comment (ranjithkumar007) [inline symengine/sets.cpp]:
```
eqn = sin(x);
soln = solve(eqn, x);
auto req = set_union({imageset(n, add(mul({i2, n, pi}), pi),
                   interval(NegInf, Inf, true, true)),
          imageset(n, mul({i2, n, pi}), interval(NegInf, Inf, true, true))});
REQUIRE(eq(*soln, *req));
```
But, then I won't be able to test something like this^.

Comment (ranjithkumar007) [inline symengine/sets.cpp]: can I just do printing test for the above like
`soln->__str__() == "2*n*pi...."` ?

Comment (isuruf) [inline symengine/sets.cpp]: Can you print the two expressions and post here?

Comment (ranjithkumar007) [inline symengine/sets.cpp]:
```
soln->__str__() = {2*pi*_n + pi | _n in (-oo, oo)} U {2*pi*_n | _n in (-oo, oo)}
req->__str__() = {2*pi*_n + pi | _n in (-oo, oo)} U {2*pi*_n | _n in (-oo, oo)}
```
both are not equal as dummy in testing module and in solve module are different.
One possible workaround could be replacing dummy `_n` with symbol `n` in `solve`  and testing with the same in `test_solve.cpp`

Comment (isuruf) [inline symengine/sets.cpp]: That's an issue with Dummy. This is not the correct place to fix it.

Comment (isuruf) [inline symengine/sets.cpp]: Okay, then don't use `Dummy`

Comment (ranjithkumar007) [inline symengine/sets.cpp]: Should I go with the method I said [above](https://github.com/symengine/symengine/pull/1314#discussion_r129218632)?

Comment (isuruf) [inline symengine/sets.cpp]: sure

Comment (ranjithkumar007) [inline symengine/sets.cpp]: Thanks for clarifying @isuruf :)

**Comment by ranjithkumar007** [comment]
@isuruf, can you review this?

**Comment by isuruf** [comment]
Can you fix the formatting as instructed in the above comment?

**Comment by ranjithkumar007** [comment]
@isuruf, It's already formatted.

File: symengine/sets.h [hunk 4]
```diff
@@ -392,6 +399,26 @@ inline RCP<const Set> imageset(const RCP<const Basic> &sym,
         return finiteset({expr});
     }
 
+    if (is_a<FiniteSet>(*base)) {
+        map_basic_basic d;
```
Comment (srajangarg) [inline symengine/sets.h]: Name this `temp`.

File: symengine/sets.h [hunk 5]
```diff
@@ -392,6 +399,26 @@ inline RCP<const Set> imageset(const RCP<const Basic> &sym,
...
+        set_basic temp;
```
Comment (srajangarg) [inline symengine/sets.h]: Name this something else. Use `[ci skip]` for next commit if these are the only changes you make.

File: symengine/subs.h [hunk 6]
```diff
@@ -169,6 +169,34 @@ class SubsVisitor : public BaseVisitor<SubsVisitor>
...
+            result_ = x.create(s, expr, bs);
+        }
+    }
+
+    void bvisit(const Union &x)
+    {
+        set_set v;
+        for (const auto &elem : x.get_container()) {
+            auto a = apply(elem);
+            if (not is_a_Set(*a))
```
Comment (srajangarg) [inline symengine/subs.h]: Not sure about this, but what if I pass `Union(x, y) -> subs({x : z})` ? It will throw, which it shouldn't right?

Comment (ranjithkumar007) [inline symengine/subs.h]: `x`, 'y' and `z` are sets?

File: symengine/sets.h (hunk 4 above)
Comment (ranjithkumar007) [inline symengine/sets.h]: Why this change?

File: symengine/subs.h (hunk 6 above)
Comment (srajangarg) [inline symengine/subs.h]: No, symbols.

Comment (ranjithkumar007) [inline symengine/subs.h]: If x,y are symbols, then Union(x,y) is FiniteSet(x,y) for which subs are not implemented yet.

File: symengine/sets.h (hunk 4 above)
Comment (srajangarg) [inline symengine/sets.h]: Okay, you can keep it as is.

File: symengine/subs.h (hunk 6 above)
Comment (srajangarg) [inline symengine/subs.h]: Can you write `subs` for FiniteSet?

**Comment by ranjithkumar007** [comment]
ping @srajangarg

**Comment by ranjithkumar007** [comment]
@isuruf can you  have a look?

**Comment by ranjithkumar007** [comment]
ping @srajangarg

**Comment by srajangarg** [comment]
I'll let @isuru merge

**Comment by ranjithkumar007** [comment]
@isuruf ping

**Comment by ranjithkumar007** [comment]
Thanks !!
